### ISSUE TITLE
Windows arm64 build

### ISSUE THREAD

**Issue by Pharoh-Ramses** [zen-browser/desktop#1051]
### What happened?

Not sure if this has been raised, but are there any plans for an arm64 build for windows?

### Reproducible?

- [X] I have checked that this issue cannot be reproduced on Mozilla Firefox.

### Version

1.0.0-a.32

### Severity impact

High

### What platform are you seeing the problem on?

Windows

**Comment by meguroyama** [comment]
+1 on ARM64 build and since FF has a build for it too finger crossed zen version is coming too.

**Comment by ectrc** [comment]
Currently for me, Zen does not even work under emulation
![image](https://github.com/user-attachments/assets/4b5cb73c-2aa6-46b9-a643-bd0a27cdc4b5)
![image](https://github.com/user-attachments/assets/3ce12012-9ff5-49db-a82c-a0d5f72ba986)

Firefox does have an arm build
![image](https://github.com/user-attachments/assets/21fb31f4-69e7-45e7-ae5c-91363ba7307a)

Would love to see this

**Comment by Okaza03** [comment]
Wold also love to see a working version on my Arm64 laptop as i love to use Zen-Browser on my desktop

**Comment by matotubol** [comment]
Also same problem on the snapdragon x Elite, would love to see this working on ARM machines.

**Comment by sky0matic** [comment]
I started using Zen on my personnal PC and work laptop.
Wanted to get it on my Snapdragon laptop too, guess I'm gonna have to stick with Brave on my laptop for now 🫤

**Comment by rexxDZN** [comment]
SAme i want this so bad on my new arm laptop...

**Comment by HarryHeres** [comment]
Since 03/09/2024, GitHub made ARM runners available to open public. I think that should be possible to implement now. 
@mauro-balades FYI

**Comment by mauro-balades** [comment]
For windows? Where did you see that?

**Comment by HarryHeres** [comment]
> For windows? Where did you see that?

https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/

**Comment by mauro-balades** [comment]
> Currently for me, Zen does not even work under emulation ![image](https://private-user-images.githubusercontent.com/13946988/362898069-4b5cb73c-2aa6-46b9-a643-bd0a27cdc4b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU3OTY3NTEsIm5iZiI6MTcyNTc5NjQ1MSwicGF0aCI6Ii8xMzk0Njk4OC8zNjI4OTgwNjktNGI1Y2I3M2MtMmFhNi00NmI5LWE2NDMtYmQwYTI3Y2RjNGI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA4VDExNTQxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiMzI2MGZjNGY2NjdiYTZlMjdjYjE5NWYzZTcwYTNmNmNjMDQzYjhlN2FlYTlhZjYzZDdkYThkMmFhNDg4N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mNXC7boxQzgKZYwRGscOFBQw0ack-IcU_3i5PwRWUew) ![image](https://private-user-images.githubusercontent.com/13946988/362898142-3ce12012-9ff5-49db-a82c-a0d5f72ba986.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU3OTY3NTEsIm5iZiI6MTcyNTc5NjQ1MSwicGF0aCI6Ii8xMzk0Njk4OC8zNjI4OTgxNDItM2NlMTIwMTItOWZmNS00OWRiLWE4MmMtYTBkNWY3MmJhOTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA4VDExNTQxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzMTMyMzZkODczYzZlNzIzMmJlZjUwNTU0MjMyMmRiMzhmMzliOWU5MDNmNWM2NGI1Nzk3YzE1ZDU4Y2UzNTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TJxO8WixztYsbuIo4k0PzGXemv6kHYCGDoG4v127Xec)
> 
> Firefox does have an arm build ![image](https://private-user-images.githubusercontent.com/13946988/362898374-21fb31f4-69e7-45e7-ae5c-91363ba7307a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU3OTY3NTEsIm5iZiI6MTcyNTc5NjQ1MSwicGF0aCI6Ii8xMzk0Njk4OC8zNjI4OTgzNzQtMjFmYjMxZjQtNjllNy00NWU3LWFlNWMtOTEzNjNiYTczMDdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA4VDExNTQxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1NjZlODEyMWVmMWYxZTJiYzA1YmRlYmZkOGVlNzQ0MzYyYzBmYTVkYjQ2ZGQ3OTk5ZTU0ZmJmMmVhYjY5NjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qO4Tdnmn5Nm5TMQBAUISMkzSkZ0vXxsVUDHith_AdQo)
> 
> Would love to see this

use the generic version

**Comment by mauro-balades** [comment]
> > For windows? Where did you see that?
> 
> https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/

The only way would be to use third party runners, id need to check their pricing

**Comment by frauderik** [comment]
> use the generic version

This is the "generic" version. I am using Snapdragon X Elite

**Comment by ectrc** [comment]
> use the generic version
I am using the "generic" version. I am using the Snapdragon X Elite.

**Comment by EndCredits** [comment]
> The only way would be to use third party runners, id need to check their pricing

According to [Supported built host - Firefox](https://firefox-source-docs.mozilla.org/contributing/build/supported.html#tier-1-hosts-and-toolchains) . It's impossible to build firefox on Windows ARM64 runner. I guess we should consider cross compile.

**Comment by EndCredits** [comment]
Detailed description about cross compile can also be found on Firefox source documentations. Official targets configuration is [L4 at mozilla-central/browser/config/mozconfigs/win64-aarch64/common-win64](https://hg.mozilla.org/mozilla-central/file/tip/browser/config/mozconfigs/win64-aarch64/common-win64#l4) . So we might be able to cross compile Zen browser by adding win64-aarch64 target like:

```
ac_add_options --target=aarch64-pc-windows-msvc
```

Considering we're using surfer. So we might should add win64-aarch64 target to surfer platform map (Reference: https://searchfox.org/mozilla-central/source/python/mozrelease/mozrelease/platforms.py#22)

```diff
diff --git a/after.ts b/browser.ts
index 8745467..ac08636 100644
--- a/src/commands/updates/browser.ts.ts
+++ b/src/commands/updates/browser.ts.ts
@@ -8,4 +8,5 @@ const ausPlatformsMap = {
   ],
   macosArm: ['Darwin_aarch64-gcc3'],
   win64: ['WINNT_x86_64-msvc', 'WINNT_x86_64-msvc-x64'],
+  win64-aarch64: ["WINNT_aarch64-msvc-aarch64"],
 }
```

Since I'm not an expert in this field so I can't provide further help. I also can't test this cuz I don't have windows x86 device. Anyway thanks a lot for your great efforts devs.

**Comment by ruverav** [comment]
Arc already has a working build for Windows ARM64 and it seems to run even faster than on my Intel-based computers. With Arc being the most direct competitor to Zen Browser, it would be interesting to see a step forward for Zen in this regard.

**Comment by nerdyinu** [comment]
+1 for linux arm64 build also

**Comment by locksvur** [comment]
+1 for windows arm64 build, especially now that arc has it. zen technically works under emulation but its super laggy and some sites don't work

**Comment by TophC7** [comment]
Looking for this as well. The difference in stability and performance is really obvious when using zen on my desktop vs my X Elite laptop.

**Comment by harryheres** [comment]
Okay, let me check this out hopefully during this week. I only have a M-based Mac, but running a VM should suffice for testing purposes.

**Comment by Molier** [comment]
> Okay, let me check this out hopefully during this week. I only have a M-based Mac, but running a VM should suffice for testing purposes.

If you need any testers, would be happy to oblige, or I can build it as well, if you run into issues with vm or something. I`m on the aarch64 snapdragon laptop.

**Comment by locksvur** [comment]
I'd also be happy to be a tester if needed :) I have a snapdragon x plus-based surface pro 11

**Comment by BenjaminJacober** [comment]
Running the generic version through the standard windows emulation layer works on my Surface Laptop 7, but a lot of pages are so slow that they're unusable (e.g. youtube).

**Comment by hardelone** [comment]
Any news? :-) I have Snapdragon  x plus in my Lenovo ThinkBook, so I´ll be happy for amr release.

**Comment by KiefBC** [comment]
Still waiting... :)

Asus Vivobook S15 OLED Copilot Laptop. Snapdragon Elite X

**Comment by ruverav** [comment]
Also willing to be a beta tester. I have a Lenovo Yoga Slim 7X laptop.

**Comment by omove** [comment]
I don't think an ARM64 runner is needed for this, changing the build target to aarch64-pc-windows-msvc and using Firefox's --enable-bootstrap build option to cross-compile worked on my x86 Fedora machine to build the ARM64 Windows version. Running the native version does immediately feel significantly faster. On BrowserBench Speedometer 3.0 the x86 version on my Yoga Slim 7x gets 5.58 vs 14.5 with the ARM64 one.

I also changed the WIN32_REDIST_DIR path from x64 to arm64 and removed the SURFER_COMPAT if-else block that has the x86 CPU optimizations.
I was going to change WIN_UCRT_REDIST_DIR from x64 to arm64 but that Windows SDK version seems to have [a bug](https://developercommunity.visualstudio.com/t/Windows-SDK-100226212428-is-missing-A/10525038#T-N10618493) and it's missing the arm64 folder, but since the DLL is apparently [part of the OS](https://developercommunity.visualstudio.com/t/Windows-SDK-100226212428-is-missing-A/10525038#T-N10618555) as of Windows 10 and overrides any application bundled ones anyway I just removed it instead and it didn't have any issues.

**Comment by yanai-sh** [comment]
Reluctantly using Arc until a native arm build is up. Patiently waiting =) 

@omove can you ELI5?

**Comment by omove** [comment]
I made this script to do it all in a fresh VM of Fedora 40.
```
#Install dependencies
sudo dnf group install -y "C Development Tools and Libraries"
sudo dnf install -y dos2unix yasm nasm gtk2-devel python3-devel m4 libuuid-devel \
     alsa-lib-devel libcurl-devel dbus-devel libdrm-devel dbus-glib-devel gtk3-devel \
     pulseaudio-libs-devel libX11-devel libXt-devel lld llvm clang-devel cbindgen \
     libstdc++-devel npm libwbclient libwbclient-devel ca-certificates msitools rustup \
     p7zip p7zip-plugins upx ccache

#Initialize rust and add rust aarch64 toolchain
rustup-init -y
. "$HOME/.cargo/env"
rustup target add aarch64-pc-windows-msvc

#Setup Zen repo
git clone https://github.com/zen-browser/desktop.git --recurse-submodules
cd desktop
npm i
npm run init
sh ./scripts/update-en-US-packs.sh

#Make changes to the windows mozconfig
sed -i '/WIN_UCRT_REDIST_DIR/ s/^/# /' ./configs/windows/mozconfig
sed -i 's/x64\/Microsoft.VC143.CRT/arm64\/Microsoft.VC143.CRT/g' ./configs/windows/mozconfig
sed -i 's/x86_64-pc-windows-msvc/aarch64-pc-windows-msvc\nac_add_options --enable-bootstrap/g' ./configs/windows/mozconfig
sed -i '/SURFER_COMPAT/,/fi/ s/^/# /' ./configs/windows/mozconfig

#Build for Windows
export SURFER_PLATFORM="win32"
npm run build
./engine/mach package

#The portable zip browser is in desktop/engine/obj-aarch64-pc-windows-msvc/dist/
#EXE Installer is located at desktop/engine/obj-aarch64-pc-windows-msvc/dist/install/sea/
```

**Comment by rexxDZN** [comment]
> I made this script to do it all in a fresh VM of Fedora 40.

@omove 
I don't think I fully understand how this work. Why create a script to be run on another OS in order to have the windows arm version of Zen Browser? Can't we just have the script to immediately run on windows, which sounds simpler than setting up a vm, then knowing how fedora works then transferring the compiled browser package to Windows. (p.s. i am not _that_ tech savvy yet)

If that CAN be the case then could u share that version of the script :)

**Comment by sky0matic** [comment]
> > I made this script to do it all in a fresh VM of Fedora 40.
> 
> @omove I don't think I fully understand how this work. Why create a script to be run on another OS in order to have the windows arm version of Zen Browser? Can't we just have the script to immediately run on windows, which sounds simpler than setting up a vm, then knowing how fedora works then transferring the compiled browser package to Windows. (p.s. i am not _that_ tech savvy yet)
> 
> If that CAN be the case then could u share that version of the script :)

What he mentioned would be to have something the Zen team can put in a pipeline that can automatically create releases for Windows ARM64.

Usually pipelines are simpler to do in Linux and they allow to cross-compile for Windows.

If you're not that tech savvy like you said, I would wait for official builds, else you would have to build it yourself for every new update they release, because staying on one version, you won't get new security updates that they do.

**Comment by omove** [comment]
I tried replacing those x86 optimizations with some ARM options instead of just commenting the whole if-else block out and I got the speedometer score from 14.5 to 16.4.
```
if test "$SURFER_COMPAT" = "true"; then
  ac_add_options --enable-optimize="-O3 -w -ftree-vectorize"

  export CFLAGS="-mtune=cortex-a55 -march=armv8.2-a+crypto -O2 -Qvec -w"
  export CPPFLAGS="-mtune=cortex-a55 -march=armv8.2-a+crypto -O2 -Qvec -w"
  export CXXFLAGS="-flto=thin -mtune=cortex-a55 -march=armv8.2-a+crypto -O2 -Qvec -w"
  export LDFLAGS="-Wl,-O2"
  export RUSTFLAGS="-Clink-args=--icf=safe -C codegen-units=1 -Ctarget-cpu=cortex-a55"
else
  ac_add_options --enable-optimize="-O2 -ftree-vectorize -Xclang -w"
  ac_add_options --enable-wasm-simd

  export CFLAGS="-O2 -march=armv8.2-a+crypto"   
  export CPPFLAGS="-O2 -march=armv8.2-a+crypto"
  export CXXFLAGS="-O2 -flto=thin -march=armv8.2-a+crypto"
  export LDFLAGS="-Wl,-O3 -march=armv8.2-a+crypto"
  export RUSTFLAGS="-Clink-args=--icf=safe -C codegen-units=1 -Ctarget-cpu=cortex-a55"    
fi
```

**Comment by Molier** [comment]
> I tried replacing those x86 optimizations with some ARM options instead of just commenting the whole if-else block out and I got the speedometer score from 14.5 to 16.4.
> 
> ```
> if test "$SURFER_COMPAT" = "true"; then
>   ac_add_options --enable-optimize="-O3 -w -ftree-vectorize"
> 
>   export CFLAGS="-mtune=cortex-a55 -march=armv8.2-a+crypto -O2 -Qvec -w"
>   export CPPFLAGS="-mtune=cortex-a55 -march=armv8.2-a+crypto -O2 -Qvec -w"
>   export CXXFLAGS="-flto=thin -mtune=cortex-a55 -march=armv8.2-a+crypto -O2 -Qvec -w"
>   export LDFLAGS="-Wl,-O2"
>   export RUSTFLAGS="-Clink-args=--icf=safe -C codegen-units=1 -Ctarget-cpu=cortex-a55"
> else
>   ac_add_options --enable-optimize="-O2 -ftree-vectorize -Xclang -w"
>   ac_add_options --enable-wasm-simd
> 
>   export CFLAGS="-O2 -march=armv8.2-a+crypto"   
>   export CPPFLAGS="-O2 -march=armv8.2-a+crypto"
>   export CXXFLAGS="-O2 -flto=thin -march=armv8.2-a+crypto"
>   export LDFLAGS="-Wl,-O3 -march=armv8.2-a+crypto"
>   export RUSTFLAGS="-Clink-args=--icf=safe -C codegen-units=1 -Ctarget-cpu=cortex-a55"    
> fi
> ```

I have tried getting it to work on fedora WSL but couldnt get your instructions to build. But probably since it was not a whole vm... 

Will try again  if it isnt supported by the end of this week. with [the recent Arc news](https://www.reddit.com/r/ArcBrowser/comments/1gc2fxx/arc_officially_not_getting_new_features_only/) I'm more than ever looking to replace it on WOA arm64 pc. So heres hoping for support 🤞 
Or if you could perhaps for now share your compiled versions @omove , think many would appreciate.

**Comment by omove** [comment]
> Will try again if it isnt supported by the end of this week. with [the recent Arc news](https://www.reddit.com/r/ArcBrowser/comments/1gc2fxx/arc_officially_not_getting_new_features_only/) I'm more than ever looking to replace it on WOA arm64 pc. So heres hoping for support 🤞 Or if you could perhaps for now share your compiled versions @omove , think many would appreciate.

I put together a PR(#3093) over the weekend so it could be merged, here's the latest arm64 build from my Github Actions test runs if you want to try it now: https://github.com/omove/desktop/actions/runs/11893761202
